pac4j-saml has incompatible OpenSAML v2 dependency on xmltooling

Hello,

I recently [opened an issue against OpenSAML v3](https://issues.shibboleth.net/jira/browse/OSJ-152), where I was getting the following stacktrace:

```
Caused by: java.lang.ClassCastException: org.opensaml.xml.schema.impl.XSAnyBuilder cannot be cast to org.opensaml.core.xml.XMLObjectBuilder
	at org.opensaml.core.xml.config.XMLConfigurator.initializeObjectProviders(XMLConfigurator.java:238)
	at org.opensaml.core.xml.config.XMLConfigurator.load(XMLConfigurator.java:203)
	at org.opensaml.core.xml.config.XMLConfigurator.load(XMLConfigurator.java:188)
	at org.opensaml.core.xml.config.XMLConfigurator.load(XMLConfigurator.java:162)
	at org.opensaml.core.xml.config.AbstractXMLObjectProviderInitializer.init(AbstractXMLObjectProviderInitializer.java:52)
	at org.opensaml.core.xml.config.XMLObjectProviderInitializer.init(XMLObjectProviderInitializer.java:45)
	at org.opensaml.core.config.InitializationService.initialize(InitializationService.java:56)
	at org.pac4j.saml.util.Configuration.bootstrap(Configuration.java:76)
	at org.pac4j.saml.util.Configuration.<clinit>(Configuration.java:49)
	... 8 more
```

The error is due to having conflicting jar resources at the same path (two jars have the same resource file in the same path, such as default-config.xml), so I presumed that it was an OpenSAML issue that only showed up if the classloader ordered jars in a specific way. However, [according to the response to my issue](https://issues.shibboleth.net/jira/browse/OSJ-152?focusedCommentId=22022&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-22022), xmltooling is part of OpenSAML v2, and is not intended nor tested alongside OpenSAML v3.

The [pac4j-saml pom.xml has an explicit dependency on xmltooling](https://github.com/pac4j/pac4j/blob/6b090ac1bb466ca038de9ae20a44284da636d289/pac4j-saml/pom.xml#L166-L168), so it appears that needs to be removed. Is there any specific reason that xmltooling is an explicit dependency of pac4j-saml?

Note that when I remove the dependency of pac4j-saml on xmltooling, that leads to two classes getting NoClassDefFoundError on `org/opensaml/xml/util/XMLHelper`: `org.pac4j.saml.metadata.SAML2IdentityProviderMetadataResolver` and `org.pac4j.saml.metadata.SAML2MetadataGenerator`. They both appear to be using XMLHelper#nodeToString, and it appears that the method has moved in OpenSAML v3 to `net.shibboleth.utilities.java.support.xml.SerializeSupport` (with the same signature and javadoc) contained in the `net.shibboleth.utilities:java-support:7.1.1` transitive dependency of opensaml-core. So, unless I missed something, it may be a quick fix.

--- a/pac4j-saml/pom.xml
+++ b/pac4j-saml/pom.xml
@@ -163,21 +163,6 @@
             <version>${spring.extensions.version}</version>
         </dependency>
         <dependency>
-            <groupId>org.opensaml</groupId>
-            <artifactId>xmltooling</artifactId>
-            <version>1.4.4</version>
-            <exclusions>
-                <exclusion>
-                    <artifactId>joda-time</artifactId>
-                    <groupId>joda-time</groupId>
-                </exclusion>
-                <exclusion>
-                    <artifactId>xmlsec</artifactId>
-                    <groupId>org.apache.santuario</groupId>
-                </exclusion>
-            </exclusions>
-        </dependency>
-        <dependency>
             <groupId>javax.servlet</groupId>
             <artifactId>javax.servlet-api</artifactId>
             <scope>provided</scope>

--- a/pac4j-saml/src/main/java/org/pac4j/saml/metadata/SAML2IdentityProviderMetadataResolver.java
+++ b/pac4j-saml/src/main/java/org/pac4j/saml/metadata/SAML2IdentityProviderMetadataResolver.java
@@ -20,6 +20,7 @@ import net.shibboleth.utilities.java.support.component.ComponentInitializationEx
 import net.shibboleth.utilities.java.support.resolver.CriteriaSet;
 import net.shibboleth.utilities.java.support.resolver.ResolverException;
 import net.shibboleth.utilities.java.support.resource.Resource;
+import net.shibboleth.utilities.java.support.xml.SerializeSupport;
 import net.shibboleth.utilities.java.support.xml.XMLParserException;
 import org.opensaml.core.criterion.EntityIdCriterion;
 import org.opensaml.core.xml.XMLObject;
@@ -27,7 +28,6 @@ import org.opensaml.saml.metadata.resolver.MetadataResolver;
 import org.opensaml.saml.metadata.resolver.impl.DOMMetadataResolver;
 import org.opensaml.saml.saml2.metadata.EntitiesDescriptor;
 import org.opensaml.saml.saml2.metadata.EntityDescriptor;
-import org.opensaml.xml.util.XMLHelper;
 import org.pac4j.core.exception.TechnicalException;
 import org.pac4j.core.util.CommonHelper;
 import org.pac4j.saml.exceptions.SAMLException;
@@ -167,7 +167,7 @@ public class SAML2IdentityProviderMetadataResolver implements SAML2MetadataResol
     public String getMetadata() {
         if (getEntityDescriptorElement() != null
                 && getEntityDescriptorElement().getDOM() != null) {
-            return XMLHelper.nodeToString(getEntityDescriptorElement().getDOM());
+            return SerializeSupport.nodeToString(getEntityDescriptorElement().getDOM());
         }
         throw new TechnicalException("Metadata cannot be retrieved because entity descriptor is null");
     }

--- a/pac4j-saml/src/main/java/org/pac4j/saml/metadata/SAML2MetadataGenerator.java
+++ b/pac4j-saml/src/main/java/org/pac4j/saml/metadata/SAML2MetadataGenerator.java
@@ -16,6 +16,7 @@
 
 package org.pac4j.saml.metadata;
 
+import net.shibboleth.utilities.java.support.xml.SerializeSupport;
 import org.apache.commons.lang.RandomStringUtils;
 import org.joda.time.DateTime;
 import org.opensaml.core.xml.XMLObjectBuilderFactory;
@@ -35,7 +36,6 @@ import org.opensaml.saml.saml2.metadata.NameIDFormat;
 import org.opensaml.saml.saml2.metadata.SPSSODescriptor;
 import org.opensaml.saml.saml2.metadata.SingleLogoutService;
 import org.opensaml.security.credential.UsageType;
-import org.opensaml.xml.util.XMLHelper;
 import org.opensaml.xmlsec.signature.KeyInfo;
 import org.pac4j.saml.crypto.CredentialProvider;
 import org.pac4j.saml.util.Configuration;
@@ -93,7 +93,7 @@ public class SAML2MetadataGenerator implements SAMLMetadataGenerator {
     public final String getMetadata() throws Exception {
         final EntityDescriptor md = buildEntityDescriptor();
         final Element entityDescriptorElement = this.marshallerFactory.getMarshaller(md).marshall(md);
-        return XMLHelper.nodeToString(entityDescriptorElement);
+        return SerializeSupport.nodeToString(entityDescriptorElement);
     }
 
     @Override

